backup-loop.sh

```diff
@@ -1,69 +1,143 @@
 #!/bin/bash
 
+set -euo pipefail
+
+readonly backup_extension="tgz"
+: "${SRC_DIR:=/data}"
+: "${DEST_DIR:=/backups}"
+: "${BACKUP_NAME:=world}"
+: "${INITIAL_DELAY:=2m}"
+: "${INTERVAL_SEC:=24h}"
+: "${PRUNE_BACKUPS_DAYS:=7}"
+: "${RCON_PORT:=25575}"
+: "${RCON_PASSWORD:=minecraft}"
+: "${EXCLUDES:=*.jar,cache,logs}" # Comma separated list of glob(3) patterns
+
+export RCON_PORT
+export RCON_PASSWORD
+
 log() {
-  level=$1
+  if [ "$#" -lt 1 ]; then
+    echo "Wrong number of arguments passed to log function" >&2
+    return 1
+  fi
+  local level="${1}"
   shift
-  echo "$(date -Iseconds) ${level} $*"
+  (
+    # If any arguments are passed besides log level
+    if [ "$#" -ge 1 ]; then
+      # then use them as log message(s)
+      <<<"${*}" cat -
+    else
+      # otherwise read log messages from standard input
+      cat -
+    fi
+  ) | awk -v level="${level}" '{ printf("%s %s %s\n", strftime("%FT%T%z"), level, $0); fflush(); }'
 }
 
-backupSet=${BACKUP_SET}
-excludes="--exclude '*.jar'"
+find_old_backups() {
+  find "${DEST_DIR}" -maxdepth 1 -name "*.${backup_extension}" -mtime "+${PRUNE_BACKUPS_DAYS}" "${@}"
+}
 
-case $TYPE in
-  FTB|CURSEFORGE)
-    cd ${SRC_DIR}/FeedTheBeast
-    ;;
-  *)
-    cd ${SRC_DIR}
-    ;;
-esac
+retry() {
+  if [ "$#" -lt 3 ]; then
+    log ERROR "Wrong number of arguments passed to retry function"
+    return 1
+  fi
+  
+  # How many times should we retry?
+  # Value smaller than zero means infinitely
+  local retries="${1}"
+  # Time to sleep between retries
+  local interval="${2}"
+  readonly retries interval
+  shift 2
+ 
+  if (( retries < 0 )); then
+    local retries_msg="infinite"
+  else
+    local retries_msg="${retries}"
+  fi
 
-backupSet="${backupSet} ${LEVEL}"
-backupSet="${backupSet} $(find . -maxdepth 1 -name '*.properties' -o -name '*.yml' -o -name '*.yaml' -o -name '*.json')"
+  local i=-1 # -1 since we will increment it before printing
+  while (( retries >= ++i )) || [ "${retries_msg}" != "${retries}" ]; do
+    # Should we also set a (high) timeout on this command?
+    # To make sure it won't freeze
+    if eval "${@}" &>/dev/null; then
+      log INFO "Command executed successfully ${*}"
+      return 0
+    else
+      log ERROR "Unable to execute ${*} - retry ${i}/${retries_msg}. Retrying in ${interval}"
+    fi
+    # shellcheck disable=SC2086
+    sleep ${interval}
+  done
+  return 2
+}
 
-if [ -d plugins ]; then
-  backupSet="${backupSet} plugins"
-fi
 
-log INFO "waiting for rcon readiness..."
-while true; do
-  rcon-cli save-on >& /dev/null && break
+# We unfortunately can't use a here-string, as it inserts new line at the end
+readarray -td, excludes_patterns < <(printf '%s' "${EXCLUDES}")
 
-  sleep 10
+excludes=()
+for pattern in "${excludes_patterns[@]}"; do
+  excludes+=(--exclude "${pattern}")
 done
-log INFO "waiting initial delay of ${INITIAL_DELAY} seconds..."
-sleep ${INITIAL_DELAY}
 
-while true; do
-  ts=$(date -u +"%Y%m%d-%H%M%S")
+# Should we even do this?
```

Comment:
with `TYPE=FTB` all the data is re-situated under `/data/FeedTheBeast`, so I believe it's a necessary evil

Reply:
As you can see I'm completely ignoring `TYPE`. If someone wants to, they could pass the `SRC_DIR=/data/FeedTheBeast` instead. Worst case scenario of deleting this logic is that the backup will contain this folder in it's root.

Reply:
Very good point. If they know they're backing up an FTB server then they'll know to change the `SRC_DIR`. I'm good to remove it then.